src/ResourceManager/Insights/Commands.Insights/Diagnostics/SetAzureRmDiagnosticSettingCommand.cs

```diff
@@ -99,105 +120,198 @@ public class SetAzureRmDiagnosticSettingCommand : ManagementCmdletBase
+                {
+                    SetSelectedTimegrains(properties);
+                }
             }
 
-            if (!string.IsNullOrWhiteSpace(this.WorkspaceId))
+            if (this.RetentionEnabled.HasValue)
             {
-                properties.WorkspaceId = this.WorkspaceId;
+                SetRetention(properties);
             }
 
-            if (this.Categories == null && this.Timegrains == null)
+            var putParameters = CopySettings(properties);
+
+            ServiceDiagnosticSettingsResource result = this.InsightsManagementClient.ServiceDiagnosticSettings.CreateOrUpdateAsync(resourceUri: this.ResourceId, parameters: putParameters, cancellationToken: CancellationToken.None).Result;
+            WriteObject(result);
+        }
+
+        private static ServiceDiagnosticSettingsResource CopySettings(ServiceDiagnosticSettingsResource properties)
+        {
+            var putParameters = new ServiceDiagnosticSettingsResource(location: string.Empty);
+            putParameters.Logs = properties.Logs;
+            putParameters.Metrics = properties.Metrics;
+            putParameters.ServiceBusRuleId = properties.ServiceBusRuleId;
+            putParameters.StorageAccountId = properties.StorageAccountId;
+            putParameters.WorkspaceId = properties.WorkspaceId;
+            return putParameters;
+        }
+
+        private void SetRetention(ServiceDiagnosticSettingsResource properties)
+        {
+            var retentionPolicy = new RetentionPolicy
+            {
+                Enabled = this.RetentionEnabled.Value,
+                Days = this.RetentionInDays.Value
+            };
+
+            if (properties.Logs != null)
             {
-                foreach (var log in properties.Logs)
+                foreach (LogSettings logSettings in properties.Logs)
                 {
-                    log.Enabled = this.Enabled;
+                    logSettings.RetentionPolicy = retentionPolicy;
                 }
+            }
 
-                foreach (var metric in properties.Metrics)
+            if (properties.Metrics != null)
+            {
+                foreach (MetricSettings metricSettings in properties.Metrics)
                 {
-                    metric.Enabled = this.Enabled;
+                    metricSettings.RetentionPolicy = retentionPolicy;
                 }
             }
-            else
+        }
+
+        private void SetSelectedTimegrains(ServiceDiagnosticSettingsResource properties)
+        {
+            if (!this.Enabled.HasValue)
             {
-                if (this.Categories != null)
-                {
-                    foreach (string category in this.Categories)
-                    {
-                        LogSettings logSettings = properties.Logs.FirstOrDefault(x => string.Equals(x.Category, category, StringComparison.OrdinalIgnoreCase));
+                throw new ArgumentException("Parameter 'Enabled' is required by 'Timegrains' parameter.");
```

Comment:
I studied the parameter set and I didn't find a configuration that satisfies all the requirements. I decided to not use them.